source/common/http/mixed_conn_pool.cc

```diff
@@ -0,0 +1,79 @@
+#include "common/http/mixed_conn_pool.h"
+
+#include "common/http/codec_client.h"
+#include "common/http/http1/conn_pool.h"
+#include "common/http/http2/conn_pool.h"
+#include "common/http/utility.h"
+#include "common/tcp/conn_pool.h"
+
+namespace Envoy {
+namespace Http {
+
+Envoy::ConnectionPool::ActiveClientPtr HttpConnPoolImplMixed::instantiateActiveClient() {
+  return std::make_unique<Tcp::ActiveTcpClient>(*this,
+                                                Envoy::ConnectionPool::ConnPoolImplBase::host(), 1);
+}
+
+CodecClientPtr
+HttpConnPoolImplMixed::createCodecClient(Upstream::Host::CreateConnectionData& data) {
+  auto protocol =
+      protocol_ == Protocol::Http11 ? CodecClient::Type::HTTP1 : CodecClient::Type::HTTP2;
+  CodecClientPtr codec{new CodecClientProd(protocol, std::move(data.connection_),
+                                           data.host_description_, dispatcher_, random_generator_)};
+  return codec;
+}
+
+void HttpConnPoolImplMixed::onConnected(Envoy::ConnectionPool::ActiveClient& client) {
+  // onConnected is called under the stack of the Network::Connection raising
+  // the Connected event. The first time it is called, it's called for a TCP
+  // client, the TCP client is detached from the connection and discarded, and an
+  // HTTP client is associated with that connection. When the first call returns, the
+  // Network::Connection will inform the new callback (the HTTP client) that it
+  // is connected. The early return is to ignore that second call.
+  if (client.protocol() != absl::nullopt) {
+    return;
+  }
+
+  connected_ = true;
+  // If an old TLS stack does not negotiate alpn, it likely does not support
+  // HTTP/2. Fail over to HTTP/1.
+  protocol_ = Protocol::Http11;
+  auto tcp_client = dynamic_cast<Tcp::ActiveTcpClient*>(&client);
```

Comment:
For perf do you want to static cast this and then assert the dynamic cast? Also, I think I asked this before, but is it possible to not have casts here by just having an interface function that returns `nextProtocol()`?

Reply:
> We could have an interface function for "detach and return the network::connection" if you prefer that to the cast?

Personally yes but up to you.

Reply:
Bah, checked this branch back out to start on that, but of course the HTTP and HTTP2 clients can't have a function that rips off the network::connection because it's owned by the codec.  given it was optional I'm going to leave as is since I think the cast is less ugly than a "hand off if possible but mostly it isn't"